cores/esp8266/core_esp8266_i2s.c

```diff
@@ -30,147 +30,278 @@
-    i2s_slc_items[x].blocksize = SLC_BUF_LEN*4;
-    i2s_slc_items[x].buf_ptr = (uint32_t)&i2s_slc_buf_pntr[x][0];
-    i2s_slc_items[x].next_link_ptr = (int)((x<(SLC_BUF_CNT-1))?(&i2s_slc_items[x+1]):(&i2s_slc_items[0]));
+    ch->slc_buf_pntr[x] = malloc(SLC_BUF_LEN*4);
+    for (y=0; y<SLC_BUF_LEN; y++) ch->slc_buf_pntr[x][y] = 0;
+
+    ch->slc_items[x].unused = 0;
+    ch->slc_items[x].owner = 1;
+    ch->slc_items[x].eof = 1;
+    ch->slc_items[x].sub_sof = 0;
+    ch->slc_items[x].datalen = SLC_BUF_LEN * 4;
+    ch->slc_items[x].blocksize = SLC_BUF_LEN * 4;
+    ch->slc_items[x].buf_ptr = (uint32_t*)&ch->slc_buf_pntr[x][0];
+    ch->slc_items[x].next_link_ptr = (x<(SLC_BUF_CNT-1))?(&ch->slc_items[x+1]):(&ch->slc_items[0]);
+    ch->slc_items[x].myid = 0;
+  }
+}
+#if 0
+void dumprx()
+{
+for (int i=0; i<SLC_BUF_CNT; i++) {
+printf("%d: %d %d %d %d %d %d %p %p\n", i, rx->slc_items[i].myid, rx->slc_items[i].owner, rx->slc_items[i].eof, rx->slc_items[i].sub_sof, rx->slc_items[i].datalen, rx->slc_items[i].blocksize,
+rx->slc_items[i].buf_ptr, rx->slc_items[i].next_link_ptr);
+}
+}
+#endif
+
+
+static void ICACHE_FLASH_ATTR i2s_slc_begin() {
+  if (tx) {
+    _alloc_channel(tx);
+  }
+  if (rx) {
+    _alloc_channel(rx);
   }
 
   ETS_SLC_INTR_DISABLE();
   SLCC0 |= SLCRXLR | SLCTXLR;
   SLCC0 &= ~(SLCRXLR | SLCTXLR);
   SLCIC = 0xFFFFFFFF;
 
-  //Configure DMA
-  SLCC0 &= ~(SLCMM << SLCM); //clear DMA MODE
-  SLCC0 |= (1 << SLCM); //set DMA MODE to 1
-  SLCRXDC |= SLCBINR | SLCBTNR; //enable INFOR_NO_REPLACE and TOKEN_NO_REPLACE
-  SLCRXDC &= ~(SLCBRXFE | SLCBRXEM | SLCBRXFM); //disable RX_FILL, RX_EOF_MODE and RX_FILL_MODE
+  // Configure DMA
+  SLCC0 &= ~(SLCMM << SLCM); // Clear DMA MODE
+  SLCC0 |= (1 << SLCM); // Set DMA MODE to 1
+  SLCRXDC |= SLCBINR | SLCBTNR; // Enable INFOR_NO_REPLACE and TOKEN_NO_REPLACE
+  SLCRXDC &= ~(/*SLCBRXFE |*/ SLCBRXEM | SLCBRXFM); //disable RX_FILL, RX_EOF_MODE and RX_FILL_MODE
 
   //Feed DMA the 1st buffer desc addr
   //To send data to the I2S subsystem, counter-intuitively we use the RXLINK part, not the TXLINK as you might
   //expect. The TXLINK part still needs a valid DMA descriptor, even if it's unused: the DMA engine will throw
   //an error at us otherwise. Just feed it any random descriptor.
   SLCTXL &= ~(SLCTXLAM << SLCTXLA); // clear TX descriptor address
-  SLCTXL |= (uint32)&i2s_slc_items[1] << SLCTXLA; //set TX descriptor address. any random desc is OK, we don't use TX but it needs to be valid
   SLCRXL &= ~(SLCRXLAM << SLCRXLA); // clear RX descriptor address
-  SLCRXL |= (uint32)&i2s_slc_items[0] << SLCRXLA; //set RX descriptor address
+  if (!rx) {
+    SLCTXL |= (uint32)&tx->slc_items[1] << SLCTXLA; // Set fake (unused) RX descriptor address
+  } else {
+    SLCTXL |= (uint32)&rx->slc_items[0] << SLCTXLA; // Set real RX address
+  }
+  if (!tx) {
+//    SLCRXL |= (uint32)&rx->slc_items[1] << SLCRXLA; // Set fake (ununsed) TX descriptor address
```

Comment:
Is this really not needed? If so, then a comment should be added to explain.

Reply:
Same as above.  Still trying to figure out *why* it's working the way it is.